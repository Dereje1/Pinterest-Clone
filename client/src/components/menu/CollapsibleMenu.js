import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import Box from '@mui/material/Box';
import Drawer from '@mui/material/Drawer';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import MenuList from '@mui/material/MenuList';
import MenuItem from '@mui/material/MenuItem';
import HomeIcon from '@mui/icons-material/Home';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import LogoutIcon from '@mui/icons-material/Logout';
import MenuIcon from '@mui/icons-material/Menu';

const ListItemTextStyling = { marginLeft: 5, fontSize: 20, fontWeight: 'bold' };

const CollapsibleMenu = ({ pathname }) => {
  const [openDrawer, setOpenDrawer] = useState(false);

  const toggleDrawer = (event) => {
    if (event.type === 'keydown' && (event.key === 'Tab' || event.key === 'Shift')) {
      return;
    }
    setOpenDrawer(!openDrawer);
  };

  const list = () => (
    <Box
      sx={{ width: 250 }}
      role="presentation"
      onClick={toggleDrawer}
      onKeyDown={toggleDrawer}
    >
      <MenuList>
        <MenuItem component={NavLink} to="/">
          <ListItemIcon>
            <HomeIcon fontSize="medium" />
          </ListItemIcon>
          <ListItemText
            sx={{ color: pathname === '/' ? '#b20e0e' : '#5e6870' }}
            primaryTypographyProps={{ ...ListItemTextStyling }}
          >
            Home
          </ListItemText>
        </MenuItem>
        <MenuItem component={NavLink} to="/pins">
          <ListItemIcon>
            <AccountCircleIcon fontSize="medium" />
          </ListItemIcon>
          <ListItemText
            sx={{ color: pathname === '/pins' ? '#b20e0e' : '#5e6870' }}
            primaryTypographyProps={{ ...ListItemTextStyling }}
          >
            My Pins
          </ListItemText>
        </MenuItem>
        <MenuItem onClick={() => window.location.assign('auth/logout')}>
          <ListItemIcon>
            <LogoutIcon fontSize="medium" />
          </ListItemIcon>
          <ListItemText
            primaryTypographyProps={{ ...ListItemTextStyling }}
            sx={{ color: '#5e6870' }}
          >
            Logout
          </ListItemText>
        </MenuItem>
      </MenuList>
    </Box>
  );

  return (
    <div>
      <MenuIcon
        onClick={toggleDrawer}
        fontSize="large"
        sx={{ marginLeft: 1 }}
      />
      <Drawer
        anchor="top"
        open={openDrawer}
        onClose={toggleDrawer}
      >
        {list()}
      </Drawer>
    </div>
  );
};

export default CollapsibleMenu;

CollapsibleMenu.propTypes = {
  pathname: PropTypes.string.isRequired,
};
