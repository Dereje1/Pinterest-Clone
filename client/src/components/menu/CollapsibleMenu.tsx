import React, { useMemo, useState } from 'react';
import { NavLink } from 'react-router-dom';
import Box from '@mui/material/Box';
import Drawer from '@mui/material/Drawer';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import MenuList from '@mui/material/MenuList';
import MenuItem from '@mui/material/MenuItem';
import HomeIcon from '@mui/icons-material/Home';
import PeopleAltOutlinedIcon from '@mui/icons-material/PeopleAltOutlined';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import LogoutIcon from '@mui/icons-material/Logout';
import MenuIcon from '@mui/icons-material/Menu';

const ListItemTextStyling = { marginLeft: 5, fontSize: 20, fontWeight: 'bold' };

const getMenuItems = (pathname: string) => ({
  home: {
    icon: <HomeIcon fontSize="medium" />,
    to: '/',
    display: 'Home',
  },
  myPins: {
    icon: <AccountCircleIcon fontSize="medium" />,
    to: '/pins',
    display: 'My Pins',
  },
  users: {
    icon: <PeopleAltOutlinedIcon fontSize="medium" />,
    to: pathname.includes('profile') ? pathname : '/profile/63acc05f21481fa569a03b0b',
    display: 'Users',
  },
  logout: {
    icon: <LogoutIcon fontSize="medium" />,
    to: '/logout',
    display: 'Logout',
  },
});

type menuItemsObject = ReturnType<typeof getMenuItems>;

function CollapsibleMenu({ pathname }: {pathname: string}) {
  const [openDrawer, setOpenDrawer] = useState(false);

  const handleKeyDown = (event: React.KeyboardEvent) => {
    if ((event.key === 'Tab' || event.key === 'Shift')) return true;
    return false;
  };

  const toggleDrawer = (event: React.SyntheticEvent) => {
    const keyDown = event as React.KeyboardEvent;
    if (event.type === 'keydown' && handleKeyDown(keyDown)) {
      return;
    }
    setOpenDrawer(!openDrawer);
  };

  const menuItems: menuItemsObject = useMemo(
    () => getMenuItems(pathname),
    [pathname],
  );

  const list = () => (
    <Box
      sx={{ width: 250 }}
      role="presentation"
      onClick={toggleDrawer}
      onKeyDown={toggleDrawer}
    >
      <MenuList>
        {
          Object.keys(menuItems).map((menu) => {
            const menuItem = menu as keyof menuItemsObject;
            return (
              <MenuItem
                component={NavLink}
                to={menuItems[menuItem].to}
                key={menu}
              >
                <ListItemIcon>
                  {menuItems[menuItem].icon}
                </ListItemIcon>
                <ListItemText
                  sx={{ color: pathname === menuItems[menuItem].to ? '#b20e0e' : '#5e6870' }}
                  primaryTypographyProps={{ ...ListItemTextStyling }}
                >
                  {menuItems[menuItem].display}
                </ListItemText>
              </MenuItem>
            );
          })
        }
      </MenuList>
    </Box>
  );

  return (
    <div>
      <MenuIcon
        onClick={toggleDrawer}
        fontSize="large"
        sx={{ marginLeft: 1, marginRight: 1 }}
      />
      <Drawer
        anchor="top"
        open={openDrawer}
        onClose={toggleDrawer}
      >
        {list()}
      </Drawer>
    </div>
  );
}

export default CollapsibleMenu;
