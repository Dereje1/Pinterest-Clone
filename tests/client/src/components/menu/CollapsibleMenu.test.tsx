/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import CollapsibleMenu from '../../../../../client/src/components/menu/CollapsibleMenu';

describe('The collapsible menu component', () => {
  test('will render for home', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('will render for my pins', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/pins" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will go to the Admin user page if users is selected', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/" />);
    const usersLink: EnzymePropSelector = wrapper.find('ForwardRef(MenuItem)').at(2);
    expect(usersLink.props().to).toBe('/profile/63acc05f21481fa569a03b0b');
  });

  test('will stay on the same page if users is selected again while on a profile page', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/some_profile" />);
    const usersLink: EnzymePropSelector = wrapper.find('ForwardRef(MenuItem)').at(2);
    expect(usersLink.props().to).toBe('/some_profile');
  });

  test('will logout the user', () => {
    const windowSpy = jest.spyOn(window, 'location', 'get');
    const mockedAssign = jest.fn();
    windowSpy.mockReturnValue({
      ...window.location,
      assign: mockedAssign,
    });
    const wrapper = shallow(<CollapsibleMenu pathname="/" />);
    const logoutLink: EnzymePropSelector = wrapper.find('ForwardRef(MenuItem)').at(3);
    logoutLink.props().onClick();
    expect(mockedAssign).toHaveBeenCalledWith('/auth/logout');
  });

  test('will toggle the drawer', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/" />);
    const menuIcon: EnzymePropSelector = wrapper.find('Memo(ForwardRef(MenuIcon))');
    let drawer: EnzymePropSelector = wrapper.find('ForwardRef(Drawer)');
    expect(drawer.props().open).toBe(false);
    menuIcon.props().onClick({ type: '' });
    drawer = wrapper.find('ForwardRef(Drawer)');
    expect(drawer.props().open).toBe(true);
  });

  test('will not toggle the drawer on certain events', () => {
    const wrapper = shallow(<CollapsibleMenu pathname="/" />);
    const menuIcon: EnzymePropSelector = wrapper.find('Memo(ForwardRef(MenuIcon))');
    let drawer: EnzymePropSelector = wrapper.find('ForwardRef(Drawer)');
    expect(drawer.props().open).toBe(false);
    menuIcon.props().onClick({ type: 'keydown', key: 'Tab' });
    drawer = wrapper.find('ForwardRef(Drawer)');
    expect(drawer.props().open).toBe(false);
    menuIcon.props().onClick({ type: 'keydown', key: 'Shift' });
    drawer = wrapper.find('ForwardRef(Drawer)');
    expect(drawer.props().open).toBe(false);
  });
});
