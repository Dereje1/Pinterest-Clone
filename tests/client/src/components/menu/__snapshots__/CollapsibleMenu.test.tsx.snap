// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The collapsible menu component will render for home 1`] = `
<div>
  <Memo(ForwardRef(MenuIcon))
    fontSize="large"
    onClick={[Function]}
    sx={
      {
        "marginLeft": 1,
        "marginRight": 1,
      }
    }
  />
  <ForwardRef(Drawer)
    anchor="top"
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(Box)
      onClick={[Function]}
      onKeyDown={[Function]}
      role="presentation"
      sx={
        {
          "width": 250,
        }
      }
    >
      <ForwardRef(MenuList)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="home"
          to="/"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(HomeIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#b20e0e",
              }
            }
          >
            Home
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="myPins"
          to="/pins"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(AccountCircleIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            My Pins
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="users"
          to="/profile/63acc05f21481fa569a03b0b"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(PeopleAltOutlinedIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            Users
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="logout"
          to="/logout"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(LogoutIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            Logout
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
      </ForwardRef(MenuList)>
    </ForwardRef(Box)>
  </ForwardRef(Drawer)>
</div>
`;

exports[`The collapsible menu component will render for my pins 1`] = `
<div>
  <Memo(ForwardRef(MenuIcon))
    fontSize="large"
    onClick={[Function]}
    sx={
      {
        "marginLeft": 1,
        "marginRight": 1,
      }
    }
  />
  <ForwardRef(Drawer)
    anchor="top"
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(Box)
      onClick={[Function]}
      onKeyDown={[Function]}
      role="presentation"
      sx={
        {
          "width": 250,
        }
      }
    >
      <ForwardRef(MenuList)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="home"
          to="/"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(HomeIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            Home
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="myPins"
          to="/pins"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(AccountCircleIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#b20e0e",
              }
            }
          >
            My Pins
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="users"
          to="/profile/63acc05f21481fa569a03b0b"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(PeopleAltOutlinedIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            Users
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
        <ForwardRef(MenuItem)
          component={
            {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          key="logout"
          to="/logout"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(LogoutIcon))
              fontSize="medium"
            />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primaryTypographyProps={
              {
                "fontSize": 20,
                "fontWeight": "bold",
                "marginLeft": 5,
              }
            }
            sx={
              {
                "color": "#5e6870",
              }
            }
          >
            Logout
          </ForwardRef(ListItemText)>
        </ForwardRef(MenuItem)>
      </ForwardRef(MenuList)>
    </ForwardRef(Box)>
  </ForwardRef(Drawer)>
</div>
`;
